io/disk/iozone.py

```diff
@@ -0,0 +1,649 @@
+from avocado.utils import build
+from avocado.utils import distro
+from avocado.utils.software_manager import SoftwareManager
+
+
+_LABELS = ['file_size', 'record_size', 'write', 'rewrite', 'read', 'reread',
+           'randread', 'randwrite', 'bkwdread', 'recordrewrite', 'strideread',
+           'fwrite', 'frewrite', 'fread', 'freread']
+
+
+def unique(list):
+    """
+    Return a list of the elements in list, but without duplicates.
+
+    :param list: List with values.
+    :return: List with non duplicate elements.
+    """
+    n = len(list)
+    if n == 0:
+        return []
+    u = {}
+    try:
+        for x in list:
+            u[x] = 1
+    except TypeError:
+        return None
+    else:
+        return u.keys()
+
+
+def geometric_mean(values):
+    """
+    Evaluates the geometric mean for a list of numeric values.
+
+    :param values: List with values.
+    :return: Single value representing the geometric mean for the list values.
+    @see: http://en.wikipedia.org/wiki/Geometric_mean
+    """
+    try:
+        values = [int(value) for value in values]
+    except ValueError:
+        return None
+    product = 1
+    n = len(values)
+    if n == 0:
+        return None
+    return math.exp(sum([math.log(x) for x in values]) / n)
+
+
+def compare_matrices(matrix1, matrix2, treshold=0.05):
+    """
+    Compare 2 matrices nxm and return a matrix nxm with comparison data
+
+    :param matrix1: Reference Matrix with numeric data
+    :param matrix2: Matrix that will be compared
+    :param treshold: Any difference bigger than this percent treshold will be
+            reported.
+    """
+    improvements = 0
+    regressions = 0
+    same = 0
+    comparison_matrix = []
+
+    new_matrix = []
+    for line1, line2 in zip(matrix1, matrix2):
+        new_line = []
+        for element1, element2 in zip(line1, line2):
+            ratio = float(element2) / float(element1)
+            if ratio < (1 - treshold):
+                regressions += 1
+                new_line.append((100 * ratio - 1) - 100)
+            elif ratio > (1 + treshold):
+                improvements += 1
+                new_line.append("+" + str((100 * ratio - 1) - 100))
+            else:
+                same + 1
+                if line1.index(element1) == 0:
+                    new_line.append(element1)
+                else:
+                    new_line.append(".")
+        new_matrix.append(new_line)
+
+    total = improvements + regressions + same
+
+    return (new_matrix, improvements, regressions, total)
+
+
+class IOzoneAnalyzer(object):
+
+    """
+    Postprocessing module for IOzone. It is capable to pick results from an
+    IOzone run, calculate the geometric mean for all throughput results for
+    a given file size or record size, and then generate a series of 2D and 3D
+    graphs. The graph generation functionality depends on gnuplot, and if it
+    is not present, functionality degrates gracefully.
+
+    @copyright: Red Hat 2010
```

Comment:
@ldoktor Oh, and the original author of the iozone wrapper was Ying Tao, but all of the analysis added later was my work and the copyright is Red Hat, since I worked for Red Hat at the time.